### PR TITLE
Performance Enhancement: Deactivate Change Detection for inactive components in exam mode

### PR DESCRIPTION
**Pull request by TobiasPr**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context

Some Students complained about performance issues during the exam. The most of them managed to work around this issue by reloading the page if the performance became too bad, but we want to improve the performance, because this is not a good a user experience.
The students reported, that the problems happend mostly with the modeling editor

### Description

- We analyzed the performance behavior of the Artemis during the exam mode, by using the chromes performance monitor.
We deactivate angular change detection for components which are inactive and also for the modeling-editor completely. This improves the performance issue, but does not entirely fix it.
- We also fixed a memory leak in the client, which was happening with the exam-timer.component. The timer was not terminated correctly, so that it still checked for the exam end even when navigating out of the exam mode.
- Another fix was made inside the apollon library to improve the performance when moving model elements.
#### Caveat: 
- We are aware, that there are some more problems with the performance. When a online code editor is loaded, the cpu usage in the exam is increased, even though the components change detection is deactivated when navigating to another exercise. This behavior indicates that there are still some processes running even though the code editor is hidden and change detection is deactivated (rxjs observable subscriptions might be the issue here).
- The CPU usage goes up when entering the exam mode (when exercise components are loaded) (on other Artemis pages 0-1%, in exam mode 3-5% on my machine). This is because of the exam timer and the periodic checks we do there. We check every 100ms if the exam is over and if we need to update the displayed time. 

#### Make sure current functionality is not broken
- create an exam (add multiple programming exercises with online-code-editor, at least 3)
- participate in it
- make sure everything works as expected
  - pay extra attention to the exam timer
  - pay extra attention to switching exercises - is the content loaded correctly?
  - pay extra attention to whether changes are synchronized correctly
  - changes in online-code-editor are detected:
      - submit something for exercise 1 and wait for the result. Are the results, build logs, tasks and problem statement correct?
      - submit something for exercise 1. While the result is still building, switch to exercise 2 and wait until the result comes in 
        (result comes in while exercise 1 is not visible). Then switch to exercise 3. Is exercise 3 still displayed correctly? Then switch to exercise 1. Are the results, build logs, tasks and problem statement correct, for exercise 2 (none) and exercise 1 correct?

For problem statement check:
- if the taskbar is correct, see screenshot
- if the tasks in the instructions are highlighted correctly, see screenshot
- if the diagram is correctly displayed and correctly highlighted, see screenshot
image

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts [hunk 1]
```diff
@@ -26,6 +26,10 @@ export class ModelingExamSubmissionComponent extends ExamSubmissionComponent imp
     exercise: ModelingExercise;
     umlModel: UMLModel; // input model for Apollon
 
+    constructor(private cdr: ChangeDetectorRef) {
```
Comment (krusche) [inline src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts:29]: please use a proper name and not `cdr`

File: src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts [hunk 2]
```diff
@@ -39,7 +43,13 @@ export class ModelingExamSubmissionComponent extends ExamSubmissionComponent imp
         return this.exercise;
     }
 
-    onActivate(): void {}
+    onActivate(): void {
+        this.cdr.reattach();
+    }
+
+    onDeactivate(): void {
+        this.cdr.reattach();
```
Comment (krusche) [inline src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts:51]: should this not be `detach()` according to https://angular.io/api/core/ChangeDetectorRef

File: src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts (hunk 1 above)
Comment (anditurdiu) [reply src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts]: See 5e32a09205c4f75d49ba9065afcb6e3699d48857

**Comment by anditurdiu** [comment]
> why are you not handling the activation and deactivation of change detection in the super class?
> This would reduce the code duplication. Please also make sure to user proper names.
> `cdr` is not a good variable name!

We moved activation and deactivation as well as ngOnInit to the superclass as most components had very similar behaviour. 
See 5e32a09205c4f75d49ba9065afcb6e3699d48857

File: src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts (hunk 2 above)
Comment (anditurdiu) [reply src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts]: See 25ab3af2b18efb7ede588c74cbb4ea22919ded30

File: src/main/webapp/app/exam/participate/exercises/exam-submission.component.ts [hunk 3]
```diff
@@ -1,14 +1,29 @@
 import { Submission } from 'app/entities/submission.model';
 import { Exercise } from 'app/entities/exercise.model';
+import { ChangeDetectorRef } from '@angular/core';
 
 export abstract class ExamSubmissionComponent {
     abstract hasUnsavedChanges(): boolean;
     abstract updateSubmissionFromView(): void;
+    abstract updateViewFromSubmission(): void;
```
Comment (schultek) [inline src/main/webapp/app/exam/participate/exercises/exam-submission.component.ts:8]: Why is this needed as an abstract method? I see it just used inside the inherited components for modeling, quiz and text exercises and not at all for programming exercises.

Comment (TobiasPr) [reply src/main/webapp/app/exam/participate/exercises/exam-submission.component.ts:8]: I think @anditurdiu added it, because it makes sense from the interface design perspective, because every ExamSubmissionComponent needs to be able to update it's content from the given submission. The code-editor is an exception, because it handles updating of the view from the submission on it's own. I see your point, that it is not needed in this case. I think it is a question on how the responsibilities of the components are in the concept. As I understood, we have the exam-participation-component, which manages all the participations and the submission components for each exercise type are just handling the submissions. Therefore the programming editor is the exception here, by handling the participation and the submission, but it is okay to add it to the interface, because future submission components very likely need to implement this method as well.

Comment (schultek) [reply src/main/webapp/app/exam/participate/exercises/exam-submission.component.ts:8]: Ok

**Review by schultek — APPROVED** [review]
Tested this on the test server with 10 exercises and **a lot** of elements in each modeling editor. I did not have any functionality or performance issues. Great work.

File: src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.ts [hunk 4]
```diff
@@ -103,11 +103,17 @@ export class ProgrammingExamSubmissionComponent extends ExamSubmissionComponent
...
     }
 
     updateSubmissionFromView(): void {
         // Note: we just save here and do not commit, because this can lead to problems!
         this.codeEditorContainer.actions.onSave();
     }
+
+    updateViewFromSubmission(): void {
+        // Note: we just save here and do not commit, because this can lead to problems!
+        this.codeEditorContainer.actions.onSave();
```
Comment (kloessst) [inline src/main/webapp/app/exam/participate/exercises/programming/programming-exam-submission.component.ts:117]: I think we could just leave this method empty and write a comment about it. Semantically it does not make sense to call `onSave` when we want to update the view from the submission (not needed for programming exercises anyways).

**Review by kloessst — APPROVED** [review]
Works well for me. I still notice some FPS drops when playing around in the modeling editor with a lot of elements but I think this can only be solved within Apollon:
PerformanceDrops

**Review by balazs-czopf — APPROVED** [review]
Tested it on a testserver, didn't have any performance problems with the modeling editor. Even with a lot of elements placed randomly into the editor everything went pretty smoothly. 
image
Didn´t encounter any other bugs or unexpected behavior.
Really nice job! 🚀

**Review by kloessst — APPROVED** [review]

**Review by balazs-czopf — APPROVED** [review]

**Review by schultek — APPROVED** [review]

**Review by schultek — APPROVED** [review]
Tested it again with the updated steps for testing and everything works fine. 👍

**Review by kloessst — APPROVED** [review]
Re-tested. Worked for me.

**Review by krusche — APPROVED** [review]
Code looks good to me

**Review by fde312 — APPROVED** [review]
Tested it with the ne testing steps and everything worked fined for me. I could not break it. Code looks good to me as well. Very good work 🚀

**Review by krusche — CHANGES_REQUESTED** [review]
@kloessst found an issue when the result is received and the corresponding code editor is in the background. Then the instructions are not displayed correctly, the same might be the case for the build logs.

I think we need to recalculate those elements based on the last result when navigating into the exercise.

**Comment by TobiasPr** [comment]
> @kloessst found an issue when the result is received and the corresponding code editor is in the background. Then the instructions are not displayed correctly, the same might be the case for the build logs.
> 
> I think we need to recalculate those elements based on the last result when navigating into the exercise.

With the help of @kloessst we added a fix for that

**Comment by TobiasPr** [comment]
There is still a problem: The task bar on top somehow displayed a wrong state - see screenshot
-> convert to draft
image

File: src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts [hunk 5]
```diff
@@ -13,6 +13,8 @@ import { Exercise } from 'app/entities/exercise.model';
     templateUrl: './modeling-exam-submission.component.html',
     providers: [{ provide: ExamSubmissionComponent, useExisting: ModelingExamSubmissionComponent }],
     styleUrls: ['./modeling-exam-submission.component.scss'],
+    // change deactivation must be triggered manually
```
Comment (pixelbrezel) [inline src/main/webapp/app/exam/participate/exercises/modeling/modeling-exam-submission.component.ts:16]: You mean `change detection must be triggered manually` right? I am a Angular noob :D, so just a question, why is the default behaviour changed here only for the modeling-exam-submission component and not for the other ones like text and so on?

**Review by pixelbrezel — COMMENTED** [review]
It worked fine on the test server and the changes seem reasonable but I lack a deep knowledge of angular change detection to approve it.
